src/khepri_machine.erl

```diff
@@ -1096,6 +1109,43 @@ get_timeout(_)                     -> khepri_app:get_default_timeout().
 clear_cache(_StoreId) ->
     ok.
 
+-define(LAST_MSG_WAS_SYNC_KEY(StoreId),
+        {?MODULE, last_message_was_sync, StoreId}).
+
+-spec last_message_was_sync(StoreId) -> ok when
+      StoreId :: khepri:store_id().
+%% @doc Remenbers that the calling process just made a synchronous command or
+%% a query.
+
+last_message_was_sync(StoreId) ->
+    Key = ?LAST_MSG_WAS_SYNC_KEY(StoreId),
+    _ = erlang:put(Key, true),
+    ok.
+
+-spec last_message_was_async(StoreId) -> ok when
+      StoreId :: khepri:store_id().
+%% @doc Remenbers that the calling process just made an asynchronous command.
+
+last_message_was_async(StoreId) ->
```

Comment:
Since we also use this function for the case where we don't know the leader or haven't sent a command yet I think it might be clearer to name this `last_message_was_not_sync/1`. To me `last_message_was_async` implies that we did send an async message which may not be the case - just that the last message wasn't sent synchronously

Reply:
I’m not satisfied with the names either.

The idea is to distinguish a sync command or a query, from an async command, an absence of any message sent to the local Ra server or a message sent to another Ra server.

Perhaps we should distinguish each of these situation:
```erlang
%% Sets the internal flag to true.
sending_query_locally(StoreId),
sending_sync_command_locally(StoreId),

%% Clears the internal flag.
sending_async_command_locally(StoreId),
sending_command_remotely(StoreId).
```

Then, the internal flag  could be named `skip_fence_preliminary_query` and the `was_last_message_sync()` renamed to `can_skip_fence_preliminary_query(StoreId)`.

Reply:
That sounds good to me :+1:

Even if it's a bit more verbose I think it's clearer about its intentions

Reply:
Cool, I will proceed with these names then. Thanks!